weave/trace_server/calls_query_builder/calls_query_builder.py

```diff
@@ -1962,3 +1966,320 @@ def _is_minimal_filter(filter: tsi.CallsFilter | None) -> bool:
+
+    # Collect values for each field across all calls
+    field_values: dict[str, list[Any]] = {
+        "ended_at": [],
+        "output_dump": [],
+        "output_refs": [],
+        "summary_dump": [],
+        "exception": [],
+        "wb_run_step_end": [],
+    }
+
+    for call in end_calls:
+        call_ids.append(call.id)
+        field_values["ended_at"].append((call.id, call.ended_at))
+        field_values["output_dump"].append((call.id, json.dumps(call.output)))
+        field_values["output_refs"].append(
+            (call.id, extract_refs_from_values(call.output))
+        )
+        field_values["summary_dump"].append((call.id, json.dumps(dict(call.summary))))
+        field_values["exception"].append((call.id, call.exception))
+        field_values["wb_run_step_end"].append((call.id, call.wb_run_step_end))
+
+    # Build optimized CASE statements (groups calls with identical values)
+    field_cases = {
+        "ended_at": _build_grouped_case_statements(
+            field_values["ended_at"], "DateTime64(6)", pb
+        ),
+        "output_dump": _build_grouped_case_statements(
+            field_values["output_dump"], "String", pb
+        ),
+        "output_refs": _build_grouped_case_statements(
+            field_values["output_refs"],
+            "Array(String)",
+            pb,
+            is_array=True,
+            array_element_type="String",
+        ),
+        "summary_dump": _build_grouped_case_statements(
+            field_values["summary_dump"], "String", pb
+        ),
+        "exception": _build_grouped_case_statements(
+            field_values["exception"], "String", pb, is_nullable=True
+        ),
+        "wb_run_step_end": _build_grouped_case_statements(
+            field_values["wb_run_step_end"], "UInt64", pb, is_nullable=True
+        ),
+    }
+
+    # Format CASE expressions with proper indentation
+    def format_cases(cases: list[str]) -> str:
+        """Format a list of CASE conditions into a multi-line string."""
+        return "\n".join(cases)
+
+    # Build WHERE clause with parameterized IN clause
+    where_params = [param_slot(pb.add_param(cid), "String") for cid in call_ids]
+    project_id_param = param_slot(pb.add_param(project_id), "String")
+
+    # Construct the final UPDATE command
+    raw_sql = f"""
+    UPDATE calls_complete
+    SET
+        ended_at = CASE {format_cases(field_cases["ended_at"])} ELSE ended_at END,
+        output_dump = CASE {format_cases(field_cases["output_dump"])} ELSE output_dump END,
+        output_refs = CASE {format_cases(field_cases["output_refs"])} ELSE output_refs END,
+        summary_dump = CASE {format_cases(field_cases["summary_dump"])} ELSE summary_dump END,
+        exception = CASE {format_cases(field_cases["exception"])} ELSE exception END,
+        wb_run_step_end = CASE {format_cases(field_cases["wb_run_step_end"])} ELSE wb_run_step_end END,
+        updated_at = now64(3)
+    WHERE project_id = {project_id_param}
```

Comment:
We can have thousands of calls per batch right? So the `CASE ...` statements could have thousands of `WHEN` parts?